include/native/stream_iteration.h

```diff
@@ -29,145 +29,179 @@
 
 namespace grppi{ 
 
-template<typename GenFunc, typename Predicate, typename OutFunc, typename ...Stages>
-void stream_iteration(parallel_execution_native &p, GenFunc && in, pipeline_info<parallel_execution_native , Stages...> && se, Predicate && condition, OutFunc && out){
-   mpmc_queue< typename std::result_of<GenFunc()>::type > queue(p.queue_size,p.lockfree);
-   mpmc_queue< typename std::result_of<GenFunc()>::type > queueOut(p.queue_size,p.lockfree);
-   std::atomic<int> nend (0);
-   std::atomic<int> nelem (0);
-   std::atomic<bool> sendFinish( false );
-   //Stream generator
-   std::thread gen([&](){
-      // Register the thread in the execution model
-      p.register_thread();
-      while(1){
-         auto k = in();
-         if(!k){
-             sendFinish=true;
-             break;
-         }
-         nelem++;
-         queue.push(k);
+template<typename Generator, typename Predicate, typename Consumer, typename ...Stages>
+void stream_iteration(parallel_execution_native &ex, Generator && generate_op, pipeline_info<parallel_execution_native , Stages...> && pipe, Predicate && predicate_op, Consumer && consume_op){
+  mpmc_queue< typename std::result_of<Generator()>::type > queue(ex.queue_size,ex.lockfree);
+  mpmc_queue< typename std::result_of<Generator()>::type > queueOut(ex.queue_size,ex.lockfree);
+  std::atomic<int> nend (0);
+  std::atomic<int> nelem (0);
+  std::atomic<bool> send_finish( false );
+  //Stream generator
+  std::thread generator_task([&](){
+    // Register the thread in the execution model
+    ex.register_thread();
+    for (;;) {
+      auto item = generate_op();
+      if (!item) {
+        send_finish=true;
+        break;
       }
-      p.deregister_thread();
-   });
-
-   std::vector<std::thread> pipeThreads;
-   composed_pipeline< mpmc_queue< typename std::result_of<GenFunc()>::type >, mpmc_queue< typename std::result_of<GenFunc()>::type >, 0, Stages ...>
-      (queue, std::forward<pipeline_info<parallel_execution_native , Stages...> >(se) , queueOut, pipeThreads); 
+      nelem++;
+      queue.push(item);
+    }
+    ex.deregister_thread();
+  });
+
+  std::vector<std::thread> pipe_threads;
```

Comment:
With 

~~~c++
using namespace std
~~~

explicit qualification is not needed